### ISSUE TITLE
cron race condition when using libnss-extrausers

### ISSUE THREAD

**Issue by dtugend** [Froxlor/Froxlor#928]
# Summary
When using libss-extrausers it can happen that other cronjobs do things that require the extra users (e.g. chown) while the extrausers file is updated by an other job.

From my understanding it also turns out that libnss-extrausers (0.6-4) has race condition problems itself when the files are changed from outside by another process (see bellow) and thus maybe should be avoided.

A short term solution would be to swtich back to libnss-mysql, but this is very slow and can stall the system a bit sporadically + high CPU load.

# System information
* Froxlor version: 0.10.24-1 (DB: 202101200)
* Web server: apache2
* DNS server: Bind
* POP/IMAP server: Dovecot
* SMTP server: postfix
* FTP server: proftpd
* OS/Version: Linux h2 4.19.0-14-amd64 #1 SMP Debian 4.19.171-2 (2021-01-30) x86_64 GNU/Linux

# Details
>  We have a server with many "customers" (90+, because I tend to separate websites by customers when they are not related) and we get these errors very randomly by email and with about an average of 1  - 2 times per week I think (not every day at least):
>
> Subject: Cron <root@h2> /usr/bin/nice -n 5 /usr/bin/php -q /var/www/froxlor/scripts/froxlor_master_cronjob.php --tasks 1> /dev/null
> Time:3/7/21, 6:40 PM
> 
> Content:
> 
> chown: invalid user: ‘froxlorlocal:froxlorlocal’
> 
> Edit: I also had it one time already where it failed with "invalid group" instead.

https://forum.froxlor.org/index.php?/topic/19710-chown-sporadically-very-randomly-failing-on-tasks-cron-job/

----

>  Same here using libnss-extrausers. So this should only happen in moments cron is running WHILE froxlor is writing new passwd/groups to /var/lib/libextrausers.
>
>But why happens froxlor jobs in parallel (writing extrausers and doing someting in scope of these users).
>
>Maybe it shoud not be:
>- deleting files
>- generating files new from SQL using a query (consuming some time)
>
>instead
>- generating files new from SQL using a query as FILE.NEW
>- delete old ones and rename/move FILE.NEW
>
>or
>- nscd may help (I removed this with migrating libnss-mysql to libnss-extrausers becauso of seldom problems nscd used 100% > CPU 

https://forum.froxlor.org/index.php?/topic/19710-chown-sporadically-very-randomly-failing-on-tasks-cron-job/&do=findComment&comment=43561

----

> I switched both servers to libnss-extrausers back then for the same reason, because libnss-mysql would make things eally slow and sometimes even stall a bit.
>
>
> This could be indeed the reason, since up to 4 tasks can run at the same time with the default /etc/cron.d/froxlor:
 >
>```
># automatically generated cron-configuration by froxlor
># do not manually edit this file as it will be re-generated periodically.
>PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
>#
>*/5 * * * * root /usr/bin/nice -n 5 /usr/bin/php -q /var/www/froxlor/scripts/froxlor_master_cronjob.php --tasks 1> /dev/null
>0 0 * * * root /usr/bin/nice -n 5 /usr/bin/php -q /var/www/froxlor/scripts/froxlor_master_cronjob.php --traffic 1> /dev/null
>5 0 * * * root /usr/bin/nice -n 5 /usr/bin/php -q /var/www/froxlor/scripts/froxlor_master_cronjob.php --usage_report 1> /dev/null
>0 */6 * * * root /usr/bin/nice -n 5 /usr/bin/php -q /var/www/froxlor/scripts/froxlor_master_cronjob.php --mailboxsize 1> /dev/null
>*/5 * * * * root /usr/bin/nice -n 5 /usr/bin/php -q /var/www/froxlor/scripts/froxlor_master_cronjob.php --letsencrypt 1> /dev/null
>10 0 * * * root /usr/bin/nice -n 5 /usr/bin/php -q /var/www/froxlor/scripts/froxlor_master_cronjob.php --backup 1> /dev/null
>```
>And each of these tasks can cause a refresh of the extrausers if more than 1 job was run:
>https://github.com/Froxlor/Froxlor/blob/8f850ee7f3c9339db0c09793496474fe6ab1f41c/lib/Froxlor/Cron/MasterCron.php#L114
>https://github.com/Froxlor/Froxlor/blob/8f850ee7f3c9339db0c09793496474fe6ab1f41c/lib/Froxlor/Cron/MasterCron.php#L131 

https://forum.froxlor.org/index.php?/topic/19710-chown-sporadically-very-randomly-failing-on-tasks-cron-job/&do=findComment&comment=43562

# Possible solutions

- Switching back to libnss-mysql, but that has quite some performance implications
- generating files more atomically, but that will not fully solve the problem, just make it less likely
- using a locking mechanism using flock (but not in PHP directly, because of it working only per process properly, meaning with multi threaded PHP it would fail / mess up), that would work if libnss-extrausers was not problematic itself (see bellow).

It _could_ work like this using the flock binary:
- Using flock binary a shared lock is obtained (directly or wrapper script)
- The lock is released and a blocking lock is obtained instead if the file needs to be updated.
- https://linux.die.net/man/1/flock
- https://serverfault.com/questions/199848/not-running-a-cron-job-when-the-prev-run-was-not-finished/609447

**Now the problem is that libnss-extrausers is very problematic itself** and should not be used in my current understanding:

It uses flockfile to get the lock on each file, now read for yourself here:
- https://packages.debian.org/source/buster/libnss-extrausers
- https://man7.org/linux/man-pages/man3/flockfile.3.html
> Note: this locking has nothing to do with the file locking done by functions like flock(2) and lockf(3).

Meaning basically it's not protected against race conditions if it's changed from outside by other processes (e.g. froxlor) as far as I understand it.
Just imagine if it accesses the file while one of the lines is still being written, so it has partial content, this could lead e.g. to the shell string part being temporarily corrupted.
While this seems unlikely, things like this happen eventually in reality usually.

In other words libnss-extrausers is not meant for the task of regularily updating the extrausers file and should be avoided maybe.

I will try to find an alternative to libnss-extrausers, but currently the only safe way I can think of is to switch back to libnss-mysql :(

**However some sort of locking or ordering meachanism for the cron jobs might not be avoidable either way, since later cronjobs might depend on the users file being correct.**

----

We could continue the discussion in the thread in the forum to avoid spamming people that follow this repository:
https://forum.froxlor.org/index.php?/topic/19710-chown-sporadically-very-randomly-failing-on-tasks-cron-job/

**Comment by d00p** [comment]
Just to be clear, your cronmail states 
`chown: invalid user: ‘froxlorlocal:froxlorlocal’ ` 

You know that this is the local user for froxlor, which has nothing to do with libnss-extrausers....

**Comment by dtugend** [comment]
Oh, that is indeed correct :S
Thank you.
Can I close the issue for now, until a better undestanding of the issue available?
